# 문제: Mongoose 메서드 반복 처리

- 서버에서 Mongoose 메서드를 사용해 ID를 배열로 받아 다 검색 후 존재 여부 판단하는 처리

```js
router.post("/deleteParts", async (req, res) => {
  // Object ID로 Site, Engine에서 쓰였는지 찾기
  console.log("id: ", req.body.id);

  let reqid = req.body.id;

  let { ok, fail } = await findQ(reqid);

  console.log("ok1: ", ok);
  console.log("fail1: ", fail);
});
```

<br><br>

## 시도

- `forEach`는 결과를 기다리지 않는다.
  - 아무리 `async/await`를 사용해도 바로 반복문 다음 코드를 실행한다.
  
  - `Array.prototype.forEach is not designed for asynchronous code.`
  - `forEach`는 각각의 Promise가 resolve 되기를 기다리지 않는다.
    - Because forEach does not wait for each promise to resolve, all the prizes are awarded in parallel, not serial (one by one).
  - 그래서, loop는 어떤 prizes도 완료되기 전에 끝나버린다.

```js
  await reqid.forEach(async (id) => {
    const cSite = await Site.find({
      partStock: { $elemMatch: { part: req.body.id } },
    });
    const cEngine = await Engine.find({
      partStock: { $elemMatch: { part: req.body.id } },
    });
    console.log(cSite);

    if (cSite && cEngine) {
      fail.push(id);
    } else {
      ok.push(id);
    }
  });
```

- `forEach`로 시도하기

```js
Array.prototype.forEachAsync = async function(cb){
    for(let x of this){
        await cb(x);
    }
}
```

<br><br>

## 해결법

- 대신해서 `for..of`를 쓰면 되지만 순차적으로 처리돼서 느려진다.
  - *하지만, 병렬적으론 처리됨??*

```js
 for( is of reqid){
    const cSite = await Site.find({
      partStock: { $elemMatch: { part: req.body.id } },
    });
    const cEngine = await Engine.find({
      partStock: { $elemMatch: { part: req.body.id } },
    });
    console.log(cSite);

    if (cSite && cEngine) {
      fail.push(id);
    } else {
      ok.push(id);
    }
});
```

<br>

- `Promise.all`을 사용해서 동시처리하면서 결과를 기다린다.
  - `.map()`을 사용해 Promise 배열을 만들어주고 `Promise.all()`을 사용해 병렬처리 후 결과를 기다려준다.

```JS
async function findQ(reqid) {
  let fail = [];
  let ok = [];

  const findPromise = await reqid.map(async (id) => {
    const cSite = await Site.find({
      partStock: { $elemMatch: { part: id } },
    });
    const cEngine = await Engine.find({
      requiredParts: { $elemMatch: { part: id } },
    });

    console.log("cSite: ", cSite);
    console.log("cEngine: ", cEngine);
    if (cSite.length === 0 && cEngine.length === 0) {
      ok.push(id);
    } else {
      fail.push(id);
    }
  });
  
  console.log("findPromise: ", findPromise);
  // findPromise:  [ Promise { <pending> }, Promise { <pending> } ] // 입력한 ID가 2개
  
  // 여기서 await가 없으면 바로 다음 코드가 실행되서 ok, fail에 담아지지 않는다.
  await Promise.all(findPromise).then((res) => {
    console.log(res);
    console.log("ok2: ", ok);
    console.log("fail2: ", fail);
  });

  return { ok, fail };
}
```

<br><br>

## Why?

- Babel을 사용하면 `async/await`가 `generator function`으로 변환되고
- `forEach`를 사용한다는 것은 각각의 iterator(반복?)가 각각의 generator 함수를 가진다는 의미이다.
  - 그래서, 각 반복이 다른 반복과 아무런 관련이 없어 독립적으로 실행되며 다른 iterator들과의 (공통된?) `next()` 컨텍스트가 없다.
  - 기본 `for` 반복문이 잘 작동하는 이유는 iterations이 하나의 generator 함수에 있기 때문이다.

`await`는 현대의 함수 평가를 중지시킨다. (모든 제어 구조를 포함하여)

<br><br><br>
<출처>

- <https://gist.github.com/joeytwiddle/37d2085425c049629b80956d3c618971>
- <https://classmethod.dev/ko/questions/37576685?page=2>
- <https://stackoverflow.com/questions/37576685/using-async-await-with-a-foreach-loop>
